### PR TITLE
add PodMigrationJob CRD proposal 

### PR DESCRIPTION
**Pull request by eahydra**
Signed-off-by: Joseph <joseph.t.lee@outlook.com>

### Ⅰ. Describe what this PR does

This proposal defines a CRD-based Pod migration API, through which the descheduler or other automatic fault recovery components can evict or delete Pods more safely. At the same time, the proposal also describes the specific implementation details of the API.

### Ⅱ. Does this pull request fix one issue?

#357

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: docs/proposals/scheduling/20220701-pod-migration-job.md [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+
+The Pod migration capability itself can be provided to users as a service. Users can integrate this API in their own systems to achieve safe migration, and are no longer limited to deschedulers.
+
+
+### Basic Migration API
+
+These APIs provide cluster administrators with more fine-grained migration control capabilities, which can better reduce risks.
+
+- Label `scheduling.koordinator.sh/refuse-to-migrate` indicates that it refuses to migrate this Pod. Although K8s assumes that all Pods should be migrated/evicted, there are always some practical exceptions, such as some stateful applications that may have serious problems if migrated, or some applications that are not yet capable of being migrated .
+- `scheduling.koordinator.sh/migration-priority` indicates the order of migration. Although the K8s community has [Pod Deletion Cost #2255](https://github.com/kubernetes/enhancements/issues/2255), it is not a general mechanism. To avoid conflicts with components that use `Pod Deletion Cost`, users can individually mark migration priorities for Pods. The higher the value, the later in the migration order.
```
Comment (hormes) [inline docs/proposals/scheduling/20220701-pod-migration-job.md:94]: These two fields may be merged, and the cost is set to max when migration is not allowed.

scheduling.koordinator.sh/eviction-cost

When we have other SLO actions, we can trigger Pod eviction, and call it eviction-cost directly?

File: docs/proposals/scheduling/20220701-pod-migration-job.md [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+	// DeleteOptions defines the deleting options for the migrated Pod and preempted Pods
+	// +optional
+	DeleteOptions *metav1.DeleteOptions `json:"deleteOptions,omitempty"`
+}
+
+type PodMigrationJobMode string
+
+const (
+	PodMigrationJobModeReservationFirst PodMigrationJobMode = "ReservationFirst"
+	PodMigrationJobModeDeleteDirectly   PodMigrationJobMode = "DeleteDirectly"
```
Comment (hormes) [inline docs/proposals/scheduling/20220701-pod-migration-job.md:147]: eviction ？

File: docs/proposals/scheduling/20220701-pod-migration-job.md [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+  - If failed to reserve, abort the PodMigrationJob and update the phase of PodMigrationJob to Fail
+  - If successfully reserve, delete the Pod through the delete method that configured in PodMigrationJob controller.
+  - Check the Reservation status to determine whether the Reservation consumed.
+  - If Reservation consumed, tracks the status of Reservation and update the status to PodMigrationJob
+  - Update phase of PodMigrationJob to Success.
+
+##### Migration Stability mechanism
+
+- Support for disabling this capability by configuration 
+- Supports a simple central flow control mechanism to limit the number of migrations over a period of time.
```
Comment (hormes) [inline docs/proposals/scheduling/20220701-pod-migration-job.md:346]: Add some description, how the user configures this security mechanism (boot parameters, configmap, or something else?)

File: docs/proposals/scheduling/20220701-pod-migration-job.md (hunk 1 above)
Comment (hormes) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:94]: For example, when the node interference is serious, and we need to evict some Pods, we hope that this value can be reused

Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:94]: Total agree.

File: docs/proposals/scheduling/20220701-pod-migration-job.md (hunk 2 above)
Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:147]: DeleteDirectly means to evict or delete the Pod directly. The exact delete behavior depends on the configuration. Or should it be described in a more abstract word like `DestroyDirectly`?

File: docs/proposals/scheduling/20220701-pod-migration-job.md (hunk 3 above)
Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md]: Added relevant details.

Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md]: Added relevant details.

File: docs/proposals/scheduling/20220701-pod-migration-job.md [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+	// Namespace represents the namespace of Reservation
+	Namespace string `json:"namespace,omitempty"`
+	// Name represents the name of Reservation
+	Name string `json:"name,omitempty"`
+	// NodeName represents the assigned node for Reservation by scheduler
+	NodeName string `json:"nodeName,omitempty"`
+	// State represents the state of Reservation
+	State string `json:"state,omitempty"`
+	// PreemptedPodRef represents the Pods that be preempted
+	PreemptedPodRef *v1.ObjectReference `json:"preemptedPodRef,omitempty"`
```
Comment (saintube) [inline docs/proposals/scheduling/20220701-pod-migration-job.md:221]: Does it mean that we can migrate at most one pod in one PodMigrationJob? It differs from the comment.

File: docs/proposals/scheduling/20220701-pod-migration-job.md [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+}
+
+type PodMigrationJobPhase string
+
+const (
+	PodMigrationJobPending PodMigrationJobPhase = "Pending"
+	PodMigrationJobRunning PodMigrationJobPhase = "Running"
+	PodMigrationJobSucceed PodMigrationJobPhase = "Succeed"
+	PodMigrationJobFailed  PodMigrationJobPhase = "Failed"
+	PodMigrationJobAborted PodMigrationJobPhase = "Aborted"
```
Comment (saintube) [inline docs/proposals/scheduling/20220701-pod-migration-job.md:251]: Could you please elaborate the usage differences between phase `Failed`, `Aborted`, `Unknown`?

File: docs/proposals/scheduling/20220701-pod-migration-job.md [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+2. Descheduling capability for different scenarios such as load-aware descheduling, defragemention, etc.
+3. The details about Deterministic preemption that preempts other Pods for Reservation.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+The descheduler in the K8s community evicts pods to be rescheduled according to different strategies. However, it does not guarantee whether the evicted Pod has resources available after re-creation. If a large number of new Pods are in the Pending state when the resources in the cluster are tight, it may lead to application availability.
```
[the remark this answers is not in the document]
Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:79]: nice!

File: docs/proposals/scheduling/20220701-pod-migration-job.md [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+
+type PodMigrationJobPreemptedReservation struct {
+	// Namespace represents the namespace of Reservation
+	Namespace string `json:"namespace,omitempty"`
+	// Name represents the name of Reservation
+	Name string `json:"name,omitempty"`
+	// NodeName represents the assigned node for Reservation by scheduler
+	NodeName string `json:"nodeName,omitempty"`
+	// State represents the state of Reservation
+	State string `json:"state,omitempty"`
```
[the remark this answers is not in the document]
Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:219]: done

File: docs/proposals/scheduling/20220701-pod-migration-job.md (hunk 4 above)
Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:221]: it's typo. fixed.

File: docs/proposals/scheduling/20220701-pod-migration-job.md (hunk 5 above)
Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:256]: done

File: docs/proposals/scheduling/20220701-pod-migration-job.md (hunk 2 above)
Comment (hormes) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:147]: Should we use to evict for this scene?

Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md]: done

File: docs/proposals/scheduling/20220701-pod-migration-job.md [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+#### Story 3
+
+The Pod migration capability itself can be provided to users as a service. Users can integrate this API in their own systems to achieve safe migration, and are no longer limited to deschedulers.
+
+
+### Basic Migration API
+
+These APIs provide cluster administrators with more fine-grained migration control capabilities, which can better reduce risks.
+
+- `scheduling.koordinator.sh/eviction-cost` indicates the cost of eviction. Although the K8s community has [Pod Deletion Cost #2255](https://github.com/kubernetes/enhancements/issues/2255), it is not a general mechanism. To avoid conflicts with components that use `Pod Deletion Cost`, users can individually mark the eviction cost for Pods. The higher the value, the later in the eviction order, and the cost is set to `math.MaxInt` when migration is not allowed.
```
Comment (hormes) [inline docs/proposals/scheduling/20220701-pod-migration-job.md:94]: Another story is how the combination of <priority, eviction-cost > takes effect. For example, whether it is the same priority, eviction-cost is considered？

Comment (hormes) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:94]: Int is 64bit in golang, the MaxInt number is too large to write. Think about MaxInt32, MaxUint32? Look at the deletion-cost constraint.

Comment (hormes) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:94]: image

File: docs/proposals/scheduling/20220701-pod-migration-job.md [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+  FlowControlBurst int32 `json:"flowControlBurst,omitempty"`
+
+  // MaxMigratingPerNode represents he maximum number of pods that can be migrating during migrate per node.
+  MaxMigratingPerNode *int32 `json:"maxMigratingPerNode,omitempty"`
+
+  // MaxMigratingPerNamespace represents he maximum number of pods that can be migrating during migrate per namespace.
+  MaxMigratingPerNamespace *int32 `json:"maxMigratingPerNamespace,omitempty"`
+
+  // MaxMigratingPerWorkload represents he maximum number of pods that can be migrating during migrate per workload.
+  MaxMigratingPerWorkload *intstr.IntOrString `json:"maxMigratingPerWorkload,omitempty"`
```
Comment (hormes) [inline docs/proposals/scheduling/20220701-pod-migration-job.md:378]: Is this value a ratio or?

File: docs/proposals/scheduling/20220701-pod-migration-job.md [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+
+  // MaxUnavailablePerWorkload represents he maximum number of pods that can be unavailable during migrate per workload.
+  // The unavailable state includes NotRunning/NotReady/Migrating/Evicting
+  MaxUnavailablePerWorkload *intstr.IntOrString `json:"maxUnavailablePerWorkload,omitempty"`
+
+  // DeleteMethod represents how to delete Pod, support "Delete" and "Eviction", default value is "Eviction"
+  DeleteMethod string `deleteMethod,omitempty`
```
Comment (hormes) [inline docs/proposals/scheduling/20220701-pod-migration-job.md:385]: Does it correspond to PodMigrationJobMode?

Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:385]: The field represents the policy of evictionDirectly..I'll change to a better name

File: docs/proposals/scheduling/20220701-pod-migration-job.md (hunk 9 above)
Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:378]: Want to support usage scenarios like 10% replicas or 5 replicas. I'll update the comment.

File: docs/proposals/scheduling/20220701-pod-migration-job.md (hunk 10 above)
Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md]: done

File: docs/proposals/scheduling/20220701-pod-migration-job.md (hunk 9 above)
Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:379]: done.

Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md:379]: done.

File: docs/proposals/scheduling/20220701-pod-migration-job.md (hunk 8 above)
Comment (eahydra) [reply docs/proposals/scheduling/20220701-pod-migration-job.md]: done.

**Pull request merged**